src/ModelContextProtocol/Configuration/McpServerBuilderExtensions.Tools.cs

```diff
@@ -55,147 +46,42 @@ public static IMcpServerBuilder WithTools(this IMcpServerBuilder builder, params
+    /// Adds types marked with the <see cref="McpServerToolTypeAttribute"/> attribute from the given assembly as tools to the server.
     /// </summary>
     /// <param name="builder">The builder instance.</param>
-    /// <param name="assembly">The assembly to load the types from. Null to get the current assembly</param>
+    /// <param name="toolAssembly">The assembly to load the types from. Null to get the current assembly</param>
     /// <exception cref="ArgumentNullException"><paramref name="builder"/> is <see langword="null"/>.</exception>
     [RequiresUnreferencedCode(RequiresUnreferencedCodeMessage)]
-    public static IMcpServerBuilder WithToolsFromAssembly(this IMcpServerBuilder builder, Assembly? assembly = null)
+    public static IMcpServerBuilder WithToolsFromAssembly(this IMcpServerBuilder builder, Assembly? toolAssembly = null)
     {
-        assembly ??= Assembly.GetCallingAssembly();
-
-        List<Type> toolTypes = [];
-
-        foreach (var type in assembly.GetTypes())
-        {
-            if (type.GetCustomAttribute<McpToolTypeAttribute>() is null)
-            {
-                continue;
-            }
+        Throw.IfNull(builder);
 
-            foreach (var method in type.GetMethods(BindingFlags.Public | BindingFlags.Static))
-            {
-                if (method.GetCustomAttribute<McpToolAttribute>() is not null)
-                {
-                    toolTypes.Add(type);
-                    break;
-                }
-            }
-        }
+        toolAssembly ??= Assembly.GetCallingAssembly();
 
-        return toolTypes.Count > 0 ?
-            WithTools(builder, toolTypes) :
-            builder;
+        return builder.WithTools(
+            from t in toolAssembly.GetTypes()
+            where t.GetCustomAttribute<McpServerToolTypeAttribute>() is not null
+            select t);
     }
 
-    private static void PopulateFunctions(
-        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type toolType, 
-        List<AIFunction> functions)
-    {
-        foreach (var method in toolType.GetMethods(BindingFlags.Public | BindingFlags.Static))
-        {
-            if (method.GetCustomAttribute<McpToolAttribute>() is not { } attribute)
-            {
-                continue;
-            }
-
-            functions.Add(AIFunctionFactory.Create(method, target: null, new()
-            {
-                Name = attribute.Name ?? method.Name,
-            }));
-        }
-    }
+    private static IEnumerable<MethodInfo> GetToolMethods(
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)] Type toolType) =>
+        from method in toolType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static)
+        where method.GetCustomAttribute<McpServerToolAttribute>() is not null
```

Comment:
Will we bring back support for non-static methods? I think people will expect it to work more like controllers at least when their using the `[McpServerToolType]`.

Reply:
Let's follow up on that separately.